src/Microsoft.AspNetCore.Blazor.Build/Core/EmbeddedResources/EmbeddedResourcesProcessor.cs

```diff
@@ -0,0 +1,137 @@
+        /// <param name="referencedAssemblyPaths">The paths to assemblies that may contain embedded resources.</param>
+        /// <param name="outputDir">The path to the directory where output is being written.</param>
+        /// <returns>A description of the embedded resources that were written to disk.</returns>
+        public static IReadOnlyList<EmbeddedResourceInfo> ExtractEmbeddedResources(
+            IEnumerable<string> referencedAssemblyPaths, string outputDir)
+        {
+            // Clean away any earlier state
+            var contentDir = Path.Combine(outputDir, ContentSubdirName);
+            if (Directory.Exists(contentDir))
+            {
+                Directory.Delete(contentDir, recursive: true);
+            }
+
+            // First, get an ordered list of AssemblyDefinition instances
+            var referencedAssemblyDefinitions = referencedAssemblyPaths
+                .Where(path => !Path.GetFileName(path).StartsWith("System.", StringComparison.Ordinal)) // Skip System.* because they are never going to contain embedded resources that we want
+                .Select(path => AssemblyDefinition.ReadAssembly(path))
+                .ToList();
+            referencedAssemblyDefinitions.Sort(OrderWithReferenceSubjectFirst);
+
+            // Now process them in turn
+            return referencedAssemblyDefinitions
+                .SelectMany(def => ExtractEmbeddedResourcesFromSingleAssembly(def, outputDir))
+                .ToList()
+                .AsReadOnly();
+        }
+
+        private static IEnumerable<EmbeddedResourceInfo> ExtractEmbeddedResourcesFromSingleAssembly(
+            AssemblyDefinition assemblyDefinition, string outputDirPath)
+        {
+            var assemblyName = assemblyDefinition.Name.Name;
+            foreach (var res in assemblyDefinition.MainModule.Resources)
+            {
+                if (TryExtractEmbeddedResource(assemblyName, res, outputDirPath, out var extractedResourceInfo))
+                {
+                    yield return extractedResourceInfo;
+                }
+            }
+        }
+
+        private static bool TryExtractEmbeddedResource(string assemblyName, Resource resource, string outputDirPath, out EmbeddedResourceInfo extractedResourceInfo)
+        {
+            if (resource is EmbeddedResource embeddedResource)
+            {
+                if (TryInterpretLogicalName(resource.Name, out var kind, out var name))
+                {
+                    // Prefix the output path with the assembly name to ensure no clashes
+                    // Also be invariant to the OS on which the package was built
+                    name = Path.Combine(ContentSubdirName, assemblyName, EnsureHasPathSeparators(name, Path.DirectorySeparatorChar));
+
+                    // Write the file content to disk, ensuring we don't try to write outside the output root
+                    var outputPath = Path.GetFullPath(Path.Combine(outputDirPath, name));
```

Comment:
In what cases could this happen? Is this just checking for the possiblity that a resource name has an absolute path (or path navigation operators) in it?

Reply:
Yes exactly. With NuGet packages (or NPM, etc) in general there’s a level of trust that they aren’t malicious, but it would be easy to envision someone accidentally making a package with an absolute pathname here and it might seem to work OK to them, but would utterly freak out the package consumer if files appeared at unrelated locations on disk.